test/test-mda-models/aissemble-test-data-delivery-pyspark-model/tests/features/environment.py

```diff
@@ -63,47 +57,104 @@ def after_scenario(context, scenario):
+
+    transit_client_token_tuple = vault.exec("cat /transit_client_token.txt")
+    transit_client_token_txt = transit_client_token_tuple.output.decode()
+    transit_client_token_json = json.loads(transit_client_token_txt)
+    encrypt_client_token = transit_client_token_json["auth"]["client_token"]
+    os.environ["ENCRYPT_CLIENT_TOKEN"] = encrypt_client_token
+
+
+def setup_s3_local(context):
+    localstack = DockerContainer("localstack/localstack:latest")
+    localstack.with_env("SERVICES", "s3")
+    port = start_container(localstack, 4566, lambda _: test_aws(localstack))
+    context.test_containers.append(localstack)
+    os.environ["S3Test_FS_PROVIDER"] = "s3"
+    os.environ["S3Test_FS_ACCESS_KEY_ID"] = "000000000000"
+    os.environ["S3Test_FS_SECRET_ACCESS_KEY"] = (
+        "E3FF2839C048B25C084DEBE9B26995E310250568"
+    )
+    os.environ["S3Test_FS_SECURE"] = "False"
+    os.environ["S3Test_FS_HOST"] = "localhost"
+    os.environ["S3Test_FS_PORT"] = f"{port}"
 
-        root_key_tuple = context.docker_compose_containers.exec_in_container(
-            ["cat", "/root_key.txt"], "vault"
-        )
-        secrets_root_key = root_key_tuple[0]
-        os.environ["SECRETS_ROOT_KEY"] = secrets_root_key
 
-        unseal_keys_tuple = context.docker_compose_containers.exec_in_container(
-            ["cat", "/unseal_keys.txt"], "vault"
-        )
-        unseal_keys_txt = unseal_keys_tuple[0]
-        unseal_keys_json = json.loads(unseal_keys_txt)
-        secrets_unseal_keys = ",".join(unseal_keys_json)
-        os.environ["SECRETS_UNSEAL_KEYS"] = secrets_unseal_keys
-
-        transit_client_token_tuple = (
-            context.docker_compose_containers.exec_in_container(
-                ["cat", "/transit_client_token.txt"], "vault"
-            )
+def after_feature(context, feature):
+    if hasattr(context, "test_containers"):
+        logger.info("Stopping Test container services")
+        for container in context.test_containers:
+            logger.info(f"...stopping {container.image}")
+            container.stop()
+
+
+def start_container(container, port, healthcheck=lambda x: True) -> int:
+    logger.info(f"Starting container: {container.image}")
+    container.with_exposed_ports(port)
+    container.start()
+    extport = container.get_exposed_port(port)
+    if not healthcheck(extport):
+        raise Exception(f"Failed to start {container.image}")
+    return extport
+
+
+def version_to_tag(version_str: str) -> str:
+    """Convert a python version into a docker tag for the same version.
+
+    Args:
+        version_str (str): The version string to convert.
+    Returns:
+        str: The docker tag for the version.
+    """
+    version = packaging.version.parse(version_str)
+    tag = version.base_version
+    if version.pre:
+        tag += "-" + "".join([str(x) for x in version.pre])
+    if version.is_devrelease:
+        tag += "-SNAPSHOT"
+    return tag
+
+
+def test_aws(container):
```

Comment:
I: I think this is less necessary now that we are confining the services started to just s3, but when we first tried this there was a connection refused in CI because it tried to authenticate before localstack was fully started.